### PR TITLE
Add drives/volumes migration from DirectCSI to DirectPV

### PR DESCRIPTION
**Pull request by cniackz**
### Objective:

To migrate the `directcsi` objects to `directpv` objects

### Details:

* The function will take the existing `directcsi` object and create new `directpv` object
* The function should be idempotent.
* Once the function succeeds, we will proceed installing the new version of directpv
* For, directcsidrives migration - we can take only the InUse and Ready drives and ignore the rest.

Screen Shot 2022-11-02 at 8 31 04 AM

### How to test:

https://github.com/cniackz/public/wiki/How-to-test-PR-652

### CODE REVIEW

**Comment by cniackz** [comment]
NEXT step: Since I can list or get the drives based on the version and status, my next step is to understand more about CRDs, how to get them or manipulate them in order to create the CRDs for the next generation of directPV. My idea is to return YAMLs for the new CRDs based on the existing ones that need to be migrated.

**Comment by cniackz** [comment]
Ok today I made some good progress, finally I fully understood where is the CRD and by the way I am already getting the old CRD into a list called filteredDrives... So good, first part is done. Now I need to create the new CRD based on the information coming from the old CRD, and for that I need first to put all my code to run in devel branch as it did in master branch. So tomorrow if time permits, I will get all needed dependencies to get my current code to work in devel branch and then I will create new CRD and in this way the task will be completed. Please give me more time and I will get this done and ask Praveen for review 

Screen Shot 2022-09-29 at 7 24 09 PM

**Comment by cniackz** [comment]
Ok at this point, my code can be compiled and executed in devel branch:

```
ccelis@asus:~/go/src/github.com/minio/directpv/migrate$ ./list-convert
 DRIVE      STATUS  VERSION                   
 /dev/dm-2  InUse   direct.csi.min.io/v1beta5 
 /dev/dm-0  InUse   direct.csi.min.io/v1beta5 
 /dev/dm-3  InUse   direct.csi.min.io/v1beta5 
 /dev/dm-1  InUse   direct.csi.min.io/v1beta5 
```

Now, I will start the convertion from `DirectCSIDrive` to `DirectPVDrive`

**Comment by cniackz** [comment]
Cool, finally I have my first converted CRD from the old CRD to the new CRD:

image

So we can use this code for the migration and additionally I am printing the new CRD in a table for inspection. Or we can use IntelliJ debugger to do so as well.

**Comment by cniackz** [comment]
I will continue with the mapping and have this done soon. Once every field is mapped and properly populated, I will request Bala and Praveen to perform the first official review of this first assigned task I have :)

**Comment by cniackz** [comment]
By the way, there are 165 files changed because I need to bring old legacy code to devel branch to get OLD crds. But once this devel branch in merged into master, I will no longer need the old* modules but whatever new name is given after all is in one place, but the code will remain pretty much the same in terms of the convertion and logic.

**Comment by cniackz** [comment]
@balamurugana and @Praveenrajmani please look into `convertCRD()` function where the convertion is taking place, we take a list of old crds and convert them to a list of new crds.

Let me know if this is what we are looking for and any feedback is appraciate it

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:65]: we could simply set `newDrive.ObjectMeta = oldDrive.ObjectMeta`

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:61]: we can use constants here, the constants should be present in `pkg/consts/consts.go`

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:55]: we have to tie it to the existing installation steps, this function will be called during the installation process and the new drives will be created

We can perhaps call this migrateDriveCRD and we need another one for volume migration

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:97]: we can use access tier types here

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:147]: we need to run discovery to find out the storage version of the object - https://github.com/minio/directpv/blob/master/pkg/client/interface.go#L79

because, `latestDirectCSIDriveInterface.List` uses dynamic clients, which will dynamically convert all the older objects to the latest ones. So the older objects will always have the `v1beta5` as its apiVersion as it comes through dynamic conversion.

**Comment by Praveenrajmani** [comment]
we only needed to copy few client API pkgs from `master`. We don't need all @cniackz

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:85]: we can use types here for states

**Review by Praveenrajmani — COMMENTED** [review]
We can add ncessary validations during conversion.
- The FSUUID should not be empty for Ready, InUse drive
- The path should be in `/dev/` notation
- The drive finalizers should be set appropriately. 

Can we also add conversion function for DirectCSIVolumes -> DirectPVVolume

**Comment by Praveenrajmani** [comment]
We can also tie this to the existing installation process @cniackz

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:65]: the name of the new drive should be its FSUUID, We need to also add validation to check if this is unique among the entire drive list set for conversion.

**Comment by cniackz** [comment]
Thank you Praveen, I will start looking into the review and feedback. Thank you!

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:65]: Yeah, make sense, no need to map field by field, will do that and test the change.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:56]: Ok, I am going to rename it `migrateDriveCRD` then. And I will also add the function for the volumes.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:113]: Ok, will do, I will use the access tier types here

Comment (cniackz) [reply migrate/list-convert.go]: Done, I am using access tier types to cast in single line all types.

**Comment by cniackz** [comment]
5 more comments to address and 1 conflict to resolve, will continue working on this today. Plan to ask review either today or next Monday morning.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:182]: Ok, I am now using `GetGroupKindVersions()` to get the values. Hope this is what is needed.

**Comment by cniackz** [comment]
4 more comments to address and 1 conflict to resolve. Still work in progress while I do other activities... Will get there.

**Comment by cniackz** [comment]
Ok @Praveenrajmani, I will investigate exactly what client API pkgs from `master` are needed and I will be exporting only those as my next step in this PR.

**Comment by cniackz** [comment]
Ok, I have copied only needed packages. Will proceed to the next point and then I will resolve the conflict.

**Comment by cniackz** [comment]
The validation was added, but I want to verify this is working in my current DirectPV environment. Please give me some more time to perform this validation here, then I will continue and fix the conflict....

**Comment by cniackz** [comment]
The validation added is working as expected in my work environment.
Next, I will work on this point:

```
 the name of the new drive should be its FSUUID, We need to also add validation to check if this is unique among the entire drive list set for conversion.
```

**Comment by cniackz** [comment]
All changes were addressed, I just have couple of questions to Praveen and need to clean code and squash PR to use one commit only. I think I will be ready by tomorrow for the second review.

**Comment by cniackz** [comment]
ok, I am ready for the review, will ping Praveen and ask him to review this PR one more time :)

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:106]: we can filter out the drives other than Ready and InUse

```go
if oldDrive.Status.DriveStatus != directcsi.DriveStatusReady && oldDrive.Status.DriveStatus != directcsi.DriveStatusInUse {
       continue
}
```

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:126]: we should also check if FSUUID is not empty

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:204]: We don't need to run discovery for each drive, Discovery is based on Group and Kind. We can move this out of the loop

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:218]: we can support migration of objects >= v1beta3 (v2.0.9) and error out for versions < v1beta3

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:224]: We should fail the migration if the fields are invalid instead of logging and moving forward

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:251]: IMO, we don't need to print anything. We can seamlessly migrate.

File: pkg/oldapis/direct.csi.min.io/v1alpha1/deepcopy_generated.go
Comment (Praveenrajmani) [inline pkg/oldapis/direct.csi.min.io/v1alpha1/deepcopy_generated.go:22]: i think we aren't using any of the deepcopy utils. We can remove this file

File: pkg/oldapis/direct.csi.min.io/v1alpha1/types.go
Comment (Praveenrajmani) [inline pkg/oldapis/direct.csi.min.io/v1alpha1/types.go:1]: we deal with only types >= v1beta3. The utils from the pkgs < v1beta3 can be removed

File: pkg/oldclient/client.go
Comment (Praveenrajmani) [inline pkg/oldclient/client.go:1]: can we remove the unused functions from this file

File: pkg/oldclient/event.go
Comment (Praveenrajmani) [inline pkg/oldclient/event.go:1]: this file can be removed

File: pkg/oldfs/errors/errors.go
Comment (Praveenrajmani) [inline pkg/oldfs/errors/errors.go:1]: we don't need this file

File: pkg/oldfs/probe.go
Comment (Praveenrajmani) [inline pkg/oldfs/probe.go:2]: we don't need this file

File: pkg/oldfs/xfs/mkfs.go
Comment (Praveenrajmani) [inline pkg/oldfs/xfs/mkfs.go:2]: we can remove this file

File: pkg/oldfs/xfs/mkfs_linux.go
Comment (Praveenrajmani) [inline pkg/oldfs/xfs/mkfs_linux.go:12]: we can remove this file

File: pkg/oldfs/xfs/mkfs_other.go
Comment (Praveenrajmani) [inline pkg/oldfs/xfs/mkfs_other.go:12]: we can remove this file

File: pkg/oldfs/xfs/quota_linux.go
Comment (Praveenrajmani) [inline pkg/oldfs/xfs/quota_linux.go:9]: we can remove this file

File: pkg/oldfs/xfs/quota_other.go
Comment (Praveenrajmani) [inline pkg/oldfs/xfs/quota_other.go:6]: we can remove this file

File: pkg/oldmount/mount_linux.go
Comment (Praveenrajmani) [inline pkg/oldmount/mount_linux.go:6]: this pkg can be removed

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:79]: Got it, it make sense, will do that then and populate this field accordingly.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:83]: Ok, I did not know this. Will remove this line.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go]: Using constant here too.

File: migrate/list-convert.go
Comment (cniackz) [reply migrate/list-convert.go:111]: Added already:

```
		// we can filter out the drives other than Ready and InUse
		if OldDrive.Status.DriveStatus != directcsi.DriveStatusReady && OldDrive.Status.DriveStatus != directcsi.DriveStatusInUse {
			continue // skip this iteration to filter out this state.
		}
```

File: cmd/kubectl-directpv/migrate.go
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/migrate.go:247]: We can't assume that the name of the drive would necessarily be FSUUID. The older objects migrated from v1alpha and v1beta1 will not have their names as FSUUIDs

File: cmd/kubectl-directpv/migrate.go
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/migrate.go:335]: we don't need to create one yaml file for each drive. We can just create one file (drives.yaml) and write all the drive manifests there. Seperate by `\n---\n`.

File: cmd/kubectl-directpv/migrate.go
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/migrate.go:366]: same here

File: cmd/kubectl-directpv/migrate_test.go
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/migrate_test.go:16]: can we have some entries in them?

File: migrate/list-convert.go
Comment (Praveenrajmani) [inline migrate/list-convert.go:128]: i think this is a duplicate. We have this function already in `cmd/kubectl-directpv/migrate.go`

File: cmd/kubectl-directpv/migrate.go
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/migrate.go:235]: `ObjectMeta.Finalizers` requires a special treatment. The older finalizers will have `direct.csi.min.io/` as its prefix. We need to change it to `directpv.min.io/`

The same for volumes.

**Review by Praveenrajmani — APPROVED** [review]

**Pull request merged**
